src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```diff
@@ -0,0 +1,241 @@
+                    username);
+            assert (account != null);
+            final PublicKeyCredentialCreationOptions options = info.options;
+            RegistrationResult result = rp
+                    .finishRegistration(FinishRegistrationOptions.builder().request(options).response(pkc).build());
+            boolean attestationIsTrusted = result.isAttestationTrusted();
+            if (attestationIsTrusted) {
+                final Set<WebAuthnCredential> previousCredentials = account.getCredentials();
+                final WebAuthnCredential newCred = new WebAuthnCredential();
+                newCred.setCreatedOn(new Date());
+                newCred.setLastUsedOn(new Date());
+                newCred.setCredentialId(result.getKeyId().getId());
+                newCred.setPublicKeyCose(result.getPublicKeyCose());
+                newCred.setSignatureCount(result.getSignatureCount());
+                newCred.setTransports(result.getKeyId().getTransports().orElse(new TreeSet<>()));
+                newCred.setParentAccount(account);
+
+                previousCredentials.add(newCred);
+                webAuthnCredentialsRepository.save(newCred);
+                account.setCredentials(previousCredentials);
+                webAuthnUserAccountRepository.save(account);
+                activeRegistrations.remove(key);
+                return Optional.of(username);
+            }
+        } catch (Exception e) {
+            System.out.println(e);
+        }
+        return Optional.empty();
+    }
+
+    public Pair<AssertionRequest, String> startLogin(String username, String realm) {
+        WebAuthnUserAccount account = webAuthnUserAccountRepository.findByProviderAndUsername(provider, username);
+        StartAssertionOptions startAssertionOptions = StartAssertionOptions.builder()
+                .userHandle(account.getUserHandle()).timeout(TIMEOUT)
+                .userVerification(UserVerificationRequirement.REQUIRED).username(username).build();
+        AssertionRequest startAssertion = rp.startAssertion(startAssertionOptions);
+        final String key = generateNewKey();
+        activeAuthentications.put(key, startAssertion);
+        return Pair.of(startAssertion, key);
+    }
+
+    String generateNewKey() {
+        String uuid = UUID.randomUUID().toString();
+        return uuid;
+    }
+
+    // public AssertionRequest startLoginUsernameless(String sessionId) {
+    // StartAssertionOptions startAssertionOptions =
+    // StartAssertionOptions.builder().timeout(TIMEOUT)
+    // .userVerification(UserVerificationRequirement.REQUIRED).build();
+    // AssertionRequest startAssertion = rp.startAssertion(startAssertionOptions);
+    // activeAuthentications.put(sessionId, startAssertion);
+    // return startAssertion;
+    // }
+
+    /**
+     * @return the authenticated username if authentication was successful, else
+     *         null
+     */
+    public Optional<String> finishLogin(
```

Comment:
usage of optional should be limited to cases where the empty result is acceptable. On login you either succeed and return the username or you fail and throw an exception

Reply:
Addressed in 141381a06e5fec8367e654b34341a7556a451790